soh/soh/Enhancements/randomizer/3drando/shops.cpp

```diff
@@ -130,74 +152,96 @@ int GetPriceFromMax(int max) {
-    if (ctx->GetOption(RSK_SHOPSANITY_PRICES).Is(RO_SHOPSANITY_PRICE_STARTER)) {
-        return 10;
-    }
-
-    static const std::vector<int> affordablePrices = { 10, 105, 205, 505 };
-    std::vector<int> priceList;
-    uint8_t maxElements = ctx->GetOption(RSK_SHOPSANITY_PRICES).Value<uint8_t>();
-    for (int i = 0; i < maxElements; i++) {
-        priceList.push_back(affordablePrices.at(i));
-    }
-    return RandomElement(priceList);
+   switch (ctx->GetOption(priceSettings.main).Value<uint8_t>()){
+      case RO_PRICE_VANILLA: 
+         return loc->GetVanillaPrice();
+      case RO_PRICE_CHEAP_BALANCED:
+         return GetCheapBalancedPrice();
+      case RO_PRICE_BALANCED:{
+         double random = RandomDouble(); //Randomly generated probability value
+         for (size_t i = 0; i < ShopPriceProbability.size(); i++) {
+            if (random < ShopPriceProbability[i]) {
+            //The randomly generated value has surpassed the total probability up to this point, so this is the generated price
+                  return  i * 5; //i in range [0, 59], output in range [0, 295] in increments of 5
+            }
+         }
+         return 150;
+      }
+      case RO_PRICE_FIXED:
+         return (uint16_t)ctx->GetOption(priceSettings.fixedPrice).Value<uint8_t>() * 5;
+      case RO_PRICE_RANGE:{
+         uint16_t range1 = (uint16_t)ctx->GetOption(priceSettings.range1).Value<uint8_t>() * 5;
+         uint16_t range2 = (uint16_t)ctx->GetOption(priceSettings.range2).Value<uint8_t>() * 5;
+         return range1 < range2 ? Random(range1, range2+1) : Random(range2, range1+1);
+      }
+      case RO_PRICE_SET_BY_WALLET:{
+         bool isTycoon = ctx->GetOption(RSK_INCLUDE_TYCOON_WALLET).Value<bool>();
+         uint16_t noWeight = ctx->GetOption(priceSettings.noWallet).Value<uint8_t>();
+         uint16_t childWeight = ctx->GetOption(priceSettings.childWallet).Value<uint8_t>();
+         uint16_t adultWeight = ctx->GetOption(priceSettings.adultWallet).Value<uint8_t>();
+         uint16_t giantWeight = ctx->GetOption(priceSettings.giantWallet).Value<uint8_t>();
+         uint16_t tycoonWeight = isTycoon ? ctx->GetOption(priceSettings.tycoonWallet).Value<uint8_t>() : 0;
+         uint16_t totalWeight = noWeight + childWeight + adultWeight + giantWeight + tycoonWeight;
+         if (totalWeight == 0){ //if no weight, return from sane range
+            return Random(0, 501);
+         }
+         totalWeight = totalWeight - noWeight;
+         if (totalWeight <= 0){
+            return 0;
+         }
+         totalWeight = totalWeight - childWeight;
+         if (totalWeight <= 0){
+            return Random(1, 99);
+         }
+         totalWeight = totalWeight - adultWeight;
+         if (totalWeight <= 0){
+            return Random(100, 200);
+         }
+         totalWeight = totalWeight - giantWeight;
+         if (totalWeight <= 0){
+            return Random(201, 500);
+         }
```

Comment:
This looks like it will just pick the range from the largest wallet with a non-zero weight, rather than using a weighted random selection. Also, Random appears to use a half-open interval.
```suggestion
         uint16_t totalWeight = noWeight + childWeight + adultWeight + giantWeight + tycoonWeight;
         if (totalWeight == 0){ //if no weight, return from sane range
            return Random(0, 501);
         }
         int16_t selected = Random(1, totalWeight + 1);
         selected = selected - noWeight;
         if (selected <= 0){
            return 0;
         }
         selected = selected - childWeight;
         if (selected <= 0){
            return Random(1, 100);
         }
         selected = selected - adultWeight;
         if (selected <= 0){
            return Random(100, 201);
         }
         selected = selected - giantWeight;
         if (selected <= 0){
            return Random(201, 501);
         }
```